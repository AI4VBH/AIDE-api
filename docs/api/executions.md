# Executions

## `GET /executions/:workflow_instance_id/tasks/:execution_id/artifacts`

This returns the task output artifacts.

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| URL | `:workflow_instance_id` | The workflow instance ID |
| URL | `:execution_id` | The task execution ID |

### Responses

This endpoint returns a dictionary of filename and MINIO object key.

| Code | Description |
|------|-------------|
| `200` | Returns output artifacts |
| `404` | Workflow instance not found or task execution not found. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/executions/9df2f059-7669-4404-b02c-427510d7db04/tasks/c2fa42a0-4a96-46b0-b599-8467ca6d3f7e/artifacts'
```

### Example Response

```json
{
  "dicom-report": "9df2f059-7669-4404-b02c-427510d7db04/c2fa42a0-4a96-46b0-b599-8467ca6d3f7e/dicom-report"
}
```

---

## `GET /executions/artifact-download?key=:object_key`

This redirects the request to a pre-signed URL from MINIO enabling the downloading of file.

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| Query | `:object_key` | The object key for the artifact you are attempting to download |

### Responses

| Code | Description |
|------|-------------|
| `302` | Redirect to the pre-signed MINIO url |
| `400` | Missing `:object_key` |
| `500` | Internal server error |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/executions/artifact-download?key=9df2f059-7669-4404-b02c-427510d7db04/c2fa42a0-4a96-46b0-b599-8467ca6d3f7e/dicom-report'
```

---

## `GET /executions/:workflow_instance_id/tasks/:execution_id/metadata`

Returns the task execution metadata.

### Response

Returns a dictionary.

| Code | Description |
|------|-------------|
| `200` | Returns execution metadata |
| `404` | Workflow instance not found or task execution not found. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/executions/9df2f059-7669-4404-b02c-427510d7db04/tasks/c2fa42a0-4a96-46b0-b599-8467ca6d3f7e/metadata'
```

### Example Response

```json
{
  "metadata-1": "value",
  "metadata-2": "value"
}
```

