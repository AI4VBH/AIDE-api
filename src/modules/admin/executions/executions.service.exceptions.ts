export enum ExecutionsServiceExceptionType {
  MISSING_TASK = 1,
}

export class ExecutionsServiceException extends Error {
  constructor(type: ExecutionsServiceExceptionType, message: string) {
    super(message);

    this.type = type;
  }

  readonly type: ExecutionsServiceExceptionType;
}
