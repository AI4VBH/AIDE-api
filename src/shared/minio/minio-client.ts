import { Inject, Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Client, ItemBucketMetadata } from 'minio';
import { parseBoolean } from 'shared/util/parseBoolean';
import { Readable } from 'stream';

@Injectable()
export class MinioClient extends Client {
  constructor(@Inject(ConfigService) config: ConfigService) {
    super({
      endPoint: config.get<string>('MINIO_HOST'),
      port: parseInt(config.get('MINIO_PORT')),
      useSSL: parseBoolean(config.get('MINIO_USE_SSL')),
      accessKey: config.get<string>('MINIO_ACCESS_KEY'),
      secretKey: config.get<string>('MINIO_SECRET_KEY'),
    });

    this.bucketName = config.get<string>('MINIO_BUCKET');

    const scheme = parseBoolean(config.get('MINIO_USE_SSL')) ? 'https' : 'http';

    this.baseUrl = `${scheme}://${config.get<string>('MINIO_HOST')}:${parseInt(
      config.get('MINIO_PORT'),
    )}`;
  }

  private readonly bucketName: string;
  public readonly baseUrl: string;

  async getPresignedObjectUrl(objectName: string): Promise<string> {
    return await this.presignedGetObject(this.bucketName, objectName, 60 * 60);
  }

  async getObjectByName(objectName: string): Promise<Readable> {
    return await this.getObject(this.bucketName, objectName);
  }

  async getObjectMetadata(objectName: string): Promise<ItemBucketMetadata> {
    const { metaData } = await this.statObject(this.bucketName, objectName);
    return metaData;
  }
}

export class MinoiClientException extends Error {
  constructor(message: string, stack?: string) {
    super(message);

    this.stack = stack;
  }
}
