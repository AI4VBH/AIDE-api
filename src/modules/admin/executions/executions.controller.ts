import {
  BadRequestException,
  Controller,
  Get,
  Inject,
  Param,
  ParseUUIDPipe,
  Query,
  Res,
  UseFilters,
} from '@nestjs/common';
import { Response } from 'express';
import ExternalServerExceptionFilter from 'shared/http/external-server-exception.filter';
import { ExecutionsService } from './executions.service';

@Controller('executions')
@UseFilters(ExternalServerExceptionFilter)
export class ExecutionsController {
  @Inject(ExecutionsService)
  private readonly executionsService: ExecutionsService;

  @Get(':workflow_instance_id/tasks/:execution_id/artifacts')
  async getWorkflowInstanceArtifacts(
    @Param('workflow_instance_id') workflowInstanceId: string,
    @Param('execution_id', ParseUUIDPipe) executionId: string,
  ) {
    return await this.executionsService.getWorkflowInstanceArtifacts(
      workflowInstanceId,
      executionId,
    );
  }

  @Get('artifact-download')
  async getArtifactDownloadUrl(
    @Query('key') file: string,
    @Res() response: Response,
  ) {
    if (!file || !file.trim()) {
      throw new BadRequestException('key query value is missing');
    }

    const { contentType, stream } = await this.executionsService.getArtifact(
      file,
    );

    if (contentType) {
      response.contentType(contentType);
    }

    const piped = stream.pipe(response);

    return piped;
  }

  @Get(':workflow_instance_id/tasks/:execution_id/metadata')
  getWorkflowInstanceMetadata(
    @Param('workflow_instance_id') workflowInstanceId: string,
    @Param('execution_id', ParseUUIDPipe) executionId: string,
  ) {
    return this.executionsService.getWorkflowInstanceMetadata(
      workflowInstanceId,
      executionId,
    );
  }
}
