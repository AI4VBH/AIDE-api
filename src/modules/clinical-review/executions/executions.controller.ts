import { Controller, Get } from '@nestjs/common';
import { Executions, ExecutionsSearchResult } from './executions.interface';
import { ExecutionsService } from './executions.service';

@Controller('executions')
export class ExecutionsController {
  constructor(private readonly executionsService: ExecutionsService) {}

  @Get()
  async getModels() {
    return await this.executionsService.getExecutions();
  }
}
