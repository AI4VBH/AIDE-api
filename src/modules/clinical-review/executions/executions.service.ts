import { Injectable } from '@nestjs/common';
import { Executions, ExecutionsSearchResult } from './executions.interface';

@Injectable()
export class ExecutionsService {
  async getExecutions() {
    // console.log('searching');
    // console.log(searchResult);
    return {};
  }
}
