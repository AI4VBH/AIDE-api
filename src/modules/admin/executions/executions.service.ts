import { HttpService } from '@nestjs/axios';
import { Inject, Injectable } from '@nestjs/common';
import { lastValueFrom } from 'rxjs';
import { MinioClient, MinoiClientException } from 'shared/minio/minio-client';
import internal from 'stream';
import { MonaiWorkflowInstance } from '../workflowinstances/workflowinstances.interface';
import {
  ExecutionsServiceException,
  ExecutionsServiceExceptionType,
} from './executions.service.exceptions';

@Injectable()
export class ExecutionsService {
  @Inject(HttpService)
  private readonly httpService: HttpService;

  @Inject(MinioClient)
  private readonly minioClient: MinioClient;

  public get minioBaseUrl(): string {
    return this.minioClient.baseUrl;
  }

  async getWorkflowInstanceArtifacts(
    workflowInstanceId: string,
    executionId: string,
  ) {
    const response = await lastValueFrom(
      this.httpService.get<MonaiWorkflowInstance>(
        `/workflowinstances/${workflowInstanceId}`,
      ),
    );

    const task = (response.data?.tasks ?? []).find(
      (t) => t.execution_id === executionId,
    );

    if (!task) {
      throw new ExecutionsServiceException(
        ExecutionsServiceExceptionType.MISSING_TASK,
        'Task not found',
      );
    }

    const artifacts = task?.output_artifacts ?? {};

    return artifacts;
  }

  async getArtifactUrl(file: string) {
    try {
      const url = await this.minioClient.getPresignedObjectUrl(file);
      return url;
    } catch (exception) {
      throw new MinoiClientException(exception.message);
    }
  }

  async getArtifact(
    file: string,
  ): Promise<{ contentType: string; stream: internal.Readable }> {
    try {
      const stream = await this.minioClient.getObjectByName(file);
      const metadata = await this.minioClient.getObjectMetadata(file);

      return { stream, contentType: metadata['content-type'] };
    } catch (exception) {
      throw new MinoiClientException(exception.code);
    }
  }

  async getWorkflowInstanceMetadata(
    workflowInstanceId: string,
    executionId: string,
  ) {
    const response = await lastValueFrom(
      this.httpService.get<MonaiWorkflowInstance>(
        `/workflowinstances/${workflowInstanceId}`,
      ),
    );

    const task = (response.data?.tasks ?? []).find(
      (t) => t.execution_id === executionId,
    );

    if (!task) {
      throw new ExecutionsServiceException(
        ExecutionsServiceExceptionType.MISSING_TASK,
        'Task not found',
      );
    }

    return task?.result ?? {};
  }
}
