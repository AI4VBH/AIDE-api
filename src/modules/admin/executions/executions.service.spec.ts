/*
 * Copyright 2022 Crown Copyright
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { DeepMocked, createMock } from '@golevelup/ts-jest';
import { HttpService } from '@nestjs/axios';
import { Test, TestingModule } from '@nestjs/testing';
import { AxiosInstance } from 'axios';
import { MinioClient } from 'shared/minio/minio-client';
import { makeObservableForTest } from 'test/utilities/test-make-observable';
import { ExecutionsService } from './executions.service';
import { ExecutionsServiceException } from './executions.service.exceptions';

describe('ExecutionsService', () => {
  let axios: DeepMocked<AxiosInstance>;
  let httpService: DeepMocked<HttpService>;
  let service: ExecutionsService;
  let minioClient: DeepMocked<MinioClient>;

  beforeEach(async () => {
    axios = createMock<AxiosInstance>();
    minioClient = createMock<MinioClient>();
    httpService = createMock<HttpService>({
      axiosRef: axios,
    });

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        ExecutionsService,
        { provide: MinioClient, useValue: minioClient },
        { provide: HttpService, useValue: httpService },
      ],
    }).compile();

    service = module.get<ExecutionsService>(ExecutionsService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('getWorkflowInstanceArtifacts', () => {
    it('returns expected result with artifacts', async () => {
      const workflowInstance = {
        tasks: [
          {
            execution_id: '6db02ec8-e77f-442d-b2b8-397f4935e907',
            status: 'Suceeded',
            output_artifacts: {
              'file.ext': 'minio-object-key/file.ext',
            },
          },
        ],
      };

      axios.get.mockResolvedValue({
        status: 200,
        data: workflowInstance,
      });

      httpService.get.mockReturnValue(makeObservableForTest(axios.get));

      const result = await service.getWorkflowInstanceArtifacts(
        '77b1e52c-6771-4385-b93d-d47970866309',
        '6db02ec8-e77f-442d-b2b8-397f4935e907',
      );

      expect(httpService.get).toHaveBeenCalled();

      expect(result).toMatchSnapshot();
    });

    it('returns expected result with no artifacts', async () => {
      const workflowInstance = {
        tasks: [
          {
            execution_id: '6db02ec8-e77f-442d-b2b8-397f4935e907',
            status: 'Suceeded',
            output_artifacts: {},
          },
        ],
      };

      axios.get.mockResolvedValue({
        status: 200,
        data: workflowInstance,
      });

      httpService.get.mockReturnValue(makeObservableForTest(axios.get));

      const result = await service.getWorkflowInstanceArtifacts(
        '77b1e52c-6771-4385-b93d-d47970866309',
        '6db02ec8-e77f-442d-b2b8-397f4935e907',
      );

      expect(httpService.get).toHaveBeenCalled();

      expect(result).toMatchSnapshot();
    });

    it('throws exception when no task is found', async () => {
      const workflowInstance = {
        tasks: [
          {
            execution_id: '6db02ec8-e77f-442d-b2b8-397f4935e907',
            status: 'Suceeded',
            output_artifacts: {},
          },
        ],
      };

      axios.get.mockResolvedValue({
        status: 200,
        data: workflowInstance,
      });

      httpService.get.mockReturnValue(makeObservableForTest(axios.get));

      const action = async () =>
        await service.getWorkflowInstanceArtifacts(
          '77b1e52c-6771-4385-b93d-d47970866309',
          'f76933bf-d157-49ae-96d6-e335a029b167',
        );

      await expect(action).rejects.toThrowError(ExecutionsServiceException);
    });
  });

  describe('getWorkflowInstanceMetadata', () => {
    it('returns expected result with artifacts', async () => {
      const workflowInstance = {
        tasks: [
          {
            execution_id: '6db02ec8-e77f-442d-b2b8-397f4935e907',
            status: 'Suceeded',
            output_artifacts: {
              'file.ext': 'minio-object-key/file.ext',
            },
            execution_stats: {
              'stat-1': 'value',
            },
          },
        ],
      };

      axios.get.mockResolvedValue({
        status: 200,
        data: workflowInstance,
      });

      httpService.get.mockReturnValue(makeObservableForTest(axios.get));

      const result = await service.getWorkflowInstanceMetadata(
        '77b1e52c-6771-4385-b93d-d47970866309',
        '6db02ec8-e77f-442d-b2b8-397f4935e907',
      );

      expect(httpService.get).toHaveBeenCalled();

      expect(result).toMatchSnapshot();
    });

    it('returns expected result with no artifacts', async () => {
      const workflowInstance = {
        tasks: [
          {
            execution_id: '6db02ec8-e77f-442d-b2b8-397f4935e907',
            status: 'Suceeded',
            output_artifacts: {},
            execution_stats: {},
          },
        ],
      };

      axios.get.mockResolvedValue({
        status: 200,
        data: workflowInstance,
      });

      httpService.get.mockReturnValue(makeObservableForTest(axios.get));

      const result = await service.getWorkflowInstanceMetadata(
        '77b1e52c-6771-4385-b93d-d47970866309',
        '6db02ec8-e77f-442d-b2b8-397f4935e907',
      );

      expect(httpService.get).toHaveBeenCalled();

      expect(result).toMatchSnapshot();
    });

    it('throws exception when no task is found', async () => {
      const workflowInstance = {
        tasks: [
          {
            execution_id: '6db02ec8-e77f-442d-b2b8-397f4935e907',
            status: 'Suceeded',
            output_artifacts: {},
            execution_stats: {},
          },
        ],
      };

      axios.get.mockResolvedValue({
        status: 200,
        data: workflowInstance,
      });

      httpService.get.mockReturnValue(makeObservableForTest(axios.get));

      const action = async () =>
        await service.getWorkflowInstanceMetadata(
          '77b1e52c-6771-4385-b93d-d47970866309',
          'f76933bf-d157-49ae-96d6-e335a029b167',
        );

      await expect(action).rejects.toThrowError(ExecutionsServiceException);
    });
  });
});
