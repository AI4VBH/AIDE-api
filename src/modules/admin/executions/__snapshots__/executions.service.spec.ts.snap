// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExecutionsService getWorkflowInstanceArtifacts returns expected result with artifacts 1`] = `
{
  "file.ext": "minio-object-key/file.ext",
}
`;

exports[`ExecutionsService getWorkflowInstanceArtifacts returns expected result with no artifacts 1`] = `{}`;

exports[`ExecutionsService getWorkflowInstanceMetadata returns expected result with artifacts 1`] = `
{
  "stat-1": "value",
}
`;

exports[`ExecutionsService getWorkflowInstanceMetadata returns expected result with no artifacts 1`] = `{}`;
